cleandart/react-dart
# Add forwardRef2/memo2 which fixes jsification of Dart props

# Motivation
Dart `forwardRef` components and `memo`-wrapped components were being passed JSified props due to the use of JS component factories, when they expected to be getting non-converted Dart props.

![forwardRef-memo-bug](https://user-images.githubusercontent.com/1713967/90829214-322e1580-e2f4-11ea-9000-5bf929881c69.png)


# Solution
__tl;dr: I added `forwardRef2`/`memo2`, added a bunch of tests around refs, and fixed a bunch of issues surfaced by test failures.__

- Add `forwardRef2` and `memo2`, which return Dart component factories so they don't convert props
    - I had to make new versions because fixing the issues required return type changes that would be breaking. We can consume these new versions under the hood in over_react's memo/uiForwardRef without any breakages.
    - Remove prop de-converting logic from `forwardRef2`
        This was only a partial solution to wrapping JS components, only applying to callbacks

        Also, now that we're using Dart component factories, this logic would prevent Dart event handlers from being converted.

        So, since we'll need a more holistic solution to the problem of wrapping JS components, and it wouldn't be easy to retain this logic without breaking other behavior, it was removed.

- Update common factory tests:
    - Refactor simple ref tests use ref test case objects for consistency and to DRY up setup
    - Add tests for all Dart factories (including forwardRef2/memo2) that Dart props are passed in without being converted
        These verify that the main issue that motivated this PR (converted JS props being passed to Dart components) is fixed
    - Add tests that verify that all ref types work with forwardRef
    - Add tests that `chainRef` works with refs that come from forwardRef2 callback args (as well as ReactElement, since it experienced a similar problem)
        - This required some code changes in different places to get these passing
    - Add test cases JS ref objects are populated with the correct values

- Update event handler tests to not test de-conversion of props via `forwardRef`

# QA Instructions
- Verify unit tests pass (CI passes)
- Verify the function components in WSD that originally surfaced these issues still work (@aaronlademann-wf)
- Smoke test examples
- Perform QA instructions in over_react PR consuming these in `uiForwardRef` and `memo`: https://github.com/Workiva/over_react/pull/620
